feat: add support for flat configs (#7935)
typescript-eslint/typescript-eslint: docs/packages/TypeScript_ESLint.mdx

Reviewer: [Docs] This should be added to `sidebar.base.js` so it shows up & has a sidebar.

Reviewer: [Docs] I also think it'd make sense to mention this package on `docs/Packages.mdx`.

export default tseslint.config(
  eslint.configs.recommended,
  ...tseslint.configs.recommended,
);
```

You can also declare our plugin and parser in your config via this package. For example this config would enable our plugin, our parser, and type-aware linting with a few of our popular type-aware rules:

```js title="eslint.config.js"
// @ts-check

import tseslint from 'typescript-eslint';

export default tseslint.config({
  plugins: {
    '@typescript-eslint': tseslint.plugin,
  },
  languageOptions: {
    parser: tseslint.parser,
    parserOptions: {
      project: true,
    },
  },
  rules: {
    '@typescript-eslint/no-unsafe-argument': 'error',
    '@typescript-eslint/no-unsafe-assignment': 'error',
    '@typescript-eslint/no-unsafe-call': 'error',
    '@typescript-eslint/no-unsafe-member-access': 'error',
    '@typescript-eslint/no-unsafe-return': 'error',
  },
});
```

:::warning
We **_strongly_** recommend declaring our plugin with the namespace `@typescript-eslint` as shown above. If you use our shared configs this is the namespace that they use. This has been the standard namespace for our plugin for many years and is what users are most familiar with.

> You may, however, choose to use a different namespace if you wish - but note that currently [flat configs allow the same plugin to be registered and configured under multiple namespaces](https://github.com/eslint/eslint/discussions/17766). This means that it's possible to have a plugin's rule turned on twice under different namespaces - leading to double reporting!
> 
> If you choose your own namespace - you are responsible for ensuring that you deduplicate usages of the plugin to ensure each rule is only enabled once.

Reviewer: [Docs] Trimming down so there's less angry red text on the page:

```suggestion
You may choose a different namespace - but note that currently [flat configs allow the same plugin to be registered, configured, and duplicately reported under multiple namespaces](https://github.com/eslint/eslint/discussions/17766).
```

I think that, counter-intutively, having more text about a _"don't do this"_ case implies that the case is more justifiable.

| `config`  | A utility function for creating type-safe flat configs     |
| `configs` | [Our eslint (flat) configs](../linting/Configurations.mdx) |
| `parser`  | [Our parser](./Parser.mdx)                                 |
| `plugin`  | [Our plugin](./ESLint_Plugin.mdx)                          |

## Usage

The `tseslint.config` function is a simple utility which does a few things. First and foremost it is a strictly typed function - meaning it allows you to write type-safe configs!

The simplest usage would be:

```js title="eslint.config.js"
// @ts-check

import eslint from '@eslint/js';
import tseslint from 'typescript-eslint';

export default tseslint.config(
  eslint.configs.recommended,
  ...tseslint.configs.recommended,
);
```

You can also declare our plugin and parser in your config via this package. For example this config would enable our plugin, our parser, and type-aware linting with a few of our popular type-aware rules:

```js title="eslint.config.js"
// @ts-check

import tseslint from 'typescript-eslint';

export default tseslint.config(
  {
    plugins: {
      '@typescript-eslint': tseslint.plugin,
    },
    languageOptions: {

>       parser: tseslint.parser,,

Reviewer 2: ```suggestion
      parser: tseslint.parser,
```

Extra comma :slightly_smiling_face:

Author: weird that prettier didn't format this out-of-the-box!
